### PR TITLE
[9.x] Replace Whoops interface with generic ExceptionRenderer interface

### PR DESCRIPTION
**Pull request by AlexVanderbist**
This PR allows for an `ExceptionRenderer` to be registered that will be used to render any exceptions.

By default, the Ignition package in laravel/laravel will register Ignition as the default `ExceptionRenderer` but this also opens up the possibility for anyone to register their own exception renderers (e.g. Whoops).

The dependency on Whoops is also dropped as it currently only serves as an interface to bootstrap Ignition.

### CODE REVIEW

**Comment by AlexVanderbist** [comment]
If this gains some traction, we'll merge the following PR in facade/ignition that implements the `ExceptionRenderer` interface: https://github.com/facade/ignition/pull/388. This way the transition to the new interface is seamless.

**Comment by GrahamCampbell** [comment]
This feels like a less powerful version of something that already exists: https://github.com/GrahamCampbell/Laravel-Exceptions.

**Comment by AlexVanderbist** [comment]
@GrahamCampbell This isn't meant as an improvement or replacement for the current exception handler. It just opens up the possibility of registering your own exception renderer without having to use Whoops' interface.

**Comment by GrahamCampbell** [comment]
That is the entire purpose of that package. To have flexible custom rendering of exceptions.

**Comment by driesvints** [comment]
@GrahamCampbell I also think this would be better off in the framework? The changes Alex is proposing are minimal.

**Comment by GrahamCampbell** [comment]
Ok. What do we think about leaving the whoops renderer in the codebase, since it's only small?

**Comment by driesvints** [comment]
@GrahamCampbell Yeah I guess that's possible. @AlexVanderbist can you maybe adjust the PR so the Whoops renderer stays in the codebase? The dependency can be moved to suggest.

**Comment by AlexVanderbist** [comment]
@driesvints so just to confirm: I'll add Whoops again as a suggested dependency and re-add the Whoops render as a fallback in `renderExceptionContent` (`ExceptionRenderer` > Whoops > Symfony)?

**Comment by GrahamCampbell** [comment]
Not as a fallback. Just a class that implements that interface, so people can use it (by writing code).

**Comment by driesvints** [comment]
@AlexVanderbist what Graham said 🙂

**Comment by taylorotwell** [comment]
So in the PR's current state, what will happen if I remove Ignition from my dependencies? Will thrown exceptions get any error display at all? Will the simple one included in Symfony be used?

**Comment by AlexVanderbist** [comment]
Re-added the existing `WhoopsHandler` and added a `WhoopsExceptionRenderer` that implements the interface. Whoops already was a suggested dependency so I didn't touch that.

@taylorotwell currently it'll fall back to Symfony. If that's an issue, `WhoopsExceptionRenderer` could be registered as the default for the `ExceptionRenderer` interface.

**Comment by taylorotwell** [comment]
So - why did we add back Whoops again? 😄

**Comment by GrahamCampbell** [comment]
Given the whoops handler is only like 10 lines, it makes sense to leave it in so people can use it if they want. Seems unnecessary to rip it out to save on 10 lines.

**Comment by GrahamCampbell** [comment]
> If we add Whoops back in like you did, how does someone actually use it?

By binding the class to `Illuminate\Contracts\Foundation\ExceptionRenderer`.

File: src/Illuminate/Contracts/Foundation/ExceptionRenderer.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+interface ExceptionRenderer
+{
+    /**
+     * Renders the given exception as HTML.
+     *
+     * @param  \Throwable  $throwable
+     * @return string
+     */
+    public function render($throwable);
```
Comment (X-Coder264) [inline src/Illuminate/Contracts/Foundation/ExceptionRenderer.php:13]: Any reason why there's no Throwable typehint here?

**Comment by taylorotwell** [comment]
One downside to this PR is that the docs link in the debug screen points to the Symfony docs - while I appreciate Symfony and what they have done for PHP that might not be of much help to Laravel devs if they are trying to figure out a Laravel problem 😅 

Any ideas on what we could do?

File: src/Illuminate/Contracts/Foundation/ExceptionRenderer.php (hunk 1 above)
Comment (AlexVanderbist) [reply src/Illuminate/Contracts/Foundation/ExceptionRenderer.php:13]: I'm not sure what the rules are regarding typehints in interfaces. Some interfaces have them, other don't...

**Comment by AlexVanderbist** [comment]
The way I see it, the Symfony page is only a fallback if you've removed Ignition from the default Laravel install and not replaced it by anything else. The default Laravel installation comes with Ignition installed and will contain a link to the Laravel docs soon.

If the framework should contain a working error page the only solution I see it to either require Whoops or Ignition in there, instead of in laravel/laravel.

**Pull request merged**